core/src/main/scala/cats/instances/option.scala

```diff
@@ -25,6 +26,12 @@ trait OptionInstances extends cats.kernel.instances.OptionInstances {
 
       def combineK[A](x: Option[A], y: Option[A]): Option[A] = if (x.isDefined) x else y
 
+      override def combineAllOptionK[A](as: IterableOnce[Option[A]]): Option[Option[A]] =
+        as.iterator.find(_.isDefined)
```

Comment:
I think we should avoid a closure in library code if we can:
```scala
val it = as.iterator
if (it.hasNext) {
  while (true) {
    val o = it.next()
    if (o.isDefined) return Some(o)
    if (!it.hasNext) return Some(None)
  }
  sys.error("unreachable")
}
else {
  return None
}
```

Reply:
Just for my own edification, why do you prefer to avoid a closure? Just avoiding unnecessary allocations or something deeper?